sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/xml/XmlSuite.scala

```diff
@@ -2115,4 +2218,1557 @@ class XmlSuite extends QueryTest with SharedSparkSession {
+    val xmlCols = xmlSchemaSf.map(sf => col(sf.name))
+    val rawXmlCols = rawXmlSchemaSf.map(sf => col(sf.name))
+    (originalDf.select(xmlCols.toIndexedSeq: _*), xmlAsRawDf.select(rawXmlCols.toIndexedSeq: _*))
+  }
+
+  private def compareOriginalAndRawXmlDataFrames(
+      originalDf: Dataset[Row],
+      rawXml: Dataset[Row],
+      colNameToBeExcludedList: List[String]): Unit = {
+    val dataFrameTuple =
+      getSpecificSchemaTupleToCheckCorrectnessOfKeepInnerXmlAsRawOption(
+        originalDf,
+        rawXml,
+        colNameToBeExcludedList)
+
+    // tuple._1 --> originalDF
+    // tuple._2 --> xmlAsRawDF
+    // Comparing schemas and apply except operator.
+    // originalDf means keepInnerXmlAsRaw option is not active
+    assert(dataFrameTuple._1.schema === dataFrameTuple._2.schema)
+    assert(dataFrameTuple._1.count() !== 0)
+    val exceptResult = dataFrameTuple._1.except(dataFrameTuple._2)
+    assert(exceptResult.collect().isEmpty)
+  }
+
+  private def applyMapFunctionToDataFrame(
+      dataFrame: Dataset[Row],
+      rowTag: String): Dataset[Row] = {
+    val map = Map(
+      "nullValue" -> "",
+      "inferSchema" -> "true",
+      "rowTag" -> rowTag,
+      "keepInnerXmlAsRaw" -> "true")
+    val options = XmlOptions(map)
+
+    val xmlInferSchema =
+      new XmlInferSchema(options, spark.sessionState.conf.caseSensitiveAnalysis)
+    val inferredSchema = xmlInferSchema.infer(dataFrame.as(Encoders.STRING).rdd)
+
+    val encoder = Encoders.row(inferredSchema)
+    // I couldn't find a good solution to convert InternalRow to Row.
+    //      expressionEncoder.resolveAndBind().createDeserializer().apply(internalRow)
+    //      Above code line didn't work as expected. It threw the following exception:
+    //      "MapObjects applied with a null function.
+    //      Likely cause is failure to resolve an array expression or encoder."
+    //    val expressionEncoder = encoder.asInstanceOf[ExpressionEncoder[Row]]
+
+    val index = dataFrame.schema.fieldIndex(rowTag)
+    val staxXmlParser = new StaxXmlParser(inferredSchema, options)
+
+    dataFrame.map(row => {
+      val finalSeq: Seq[Any] = staxXmlParser
+        .parseColumn(row.getString(index), inferredSchema)
+        .toSeq(inferredSchema)
+      val arrayBuffer = new ArrayBuffer[Any]
+      for (elem <- finalSeq) {
+
+        elem match {
+          case data: GenericArrayData =>
+            val arrayBufferInner = new ArrayBuffer[Any]
+            for (elem <- data.array) {
+              if (elem.isInstanceOf[UTF8String]) {
+                arrayBufferInner.addOne(elem.toString)
+              } else {
+                arrayBufferInner.addOne(elem)
+              }
+            }
+            arrayBuffer.addOne(arrayBufferInner)
+          case data: UTF8String =>
+            arrayBuffer.addOne(data.toString)
+          case _ =>
+            arrayBuffer.addOne(elem)
+        }
+      }
+      Row.fromSeq(arrayBuffer.toSeq)
```

Comment:
I couldn't find a good solution to convert InternalRow to Row. parseColumn method returns InternalRow but I want to convert it to Row. I am looking for a more direct solution for conversion operation. Is there any suggestion for conversion InternalRow to Row